packages/langium/src/grammar/validation/validator.ts

```diff
@@ -5,31 +5,30 @@
-import { NamedAstNode } from '../references/name-provider';
-import { References } from '../references/references';
-import { LangiumServices } from '../services';
-import { AstNode, Properties, Reference } from '../syntax-tree';
-import { getContainerOfType, streamAllContents } from '../utils/ast-util';
-import { MultiMap } from '../utils/collections';
-import { toDocumentSegment } from '../utils/cst-util';
-import { findNodeForKeyword, findNameAssignment, findNodeForProperty, getAllReachableRules } from '../utils/grammar-util';
-import { Stream, stream } from '../utils/stream';
-import { ValidationAcceptor, ValidationChecks, ValidationRegistry } from '../validation/validation-registry';
-import { LangiumDocuments } from '../workspace/documents';
-import * as ast from './generated/ast';
-import { isParserRule, isRuleCall } from './generated/ast';
-import { getTypeName, isDataTypeRule, isOptionalCardinality, resolveImport, resolveTransitiveImports, terminalRegex } from './internal-grammar-util';
-import type { LangiumGrammarServices } from './langium-grammar-module';
-import { applyErrorToAssignment, collectAllInterfaces, InterfaceInfo, validateTypesConsistency } from './type-system/type-validator';
+import { NamedAstNode } from '../../references/name-provider';
+import { References } from '../../references/references';
+import { AstNode, Properties, Reference } from '../../syntax-tree';
+import { getContainerOfType, streamAllContents } from '../../utils/ast-util';
+import { MultiMap } from '../../utils/collections';
+import { toDocumentSegment } from '../../utils/cst-util';
+import { findNodeForKeyword, findNameAssignment, findNodeForProperty, getAllReachableRules } from '../../utils/grammar-util';
+import { Stream, stream } from '../../utils/stream';
+import { ValidationAcceptor, ValidationChecks, ValidationRegistry } from '../../validation/validation-registry';
+import { LangiumDocuments } from '../../workspace/documents';
+import * as ast from '../generated/ast';
+import { isParserRule, isRuleCall } from '../generated/ast';
+import { getTypeNameWithoutError, isDataTypeRule, isOptionalCardinality, isPrimitiveType, resolveImport, resolveTransitiveImports, terminalRegex } from '../internal-grammar-util';
+import type { LangiumGrammarServices } from '../langium-grammar-module';
 
 export class LangiumGrammarValidationRegistry extends ValidationRegistry {
     constructor(services: LangiumGrammarServices) {
         super(services);
         const validator = services.validation.LangiumGrammarValidator;
+        const typesValidator = services.validation.LangiumGrammarTypesValidator;
         const checks: ValidationChecks<ast.LangiumGrammarAstType> = {
             Action: [
-                validator.checkActionTypeUnions,
-                validator.checkAssignmentReservedName
+                validator.checkAssignmentReservedName,
+                typesValidator.checkActionIsNotUnionType,
```

Comment:
It is incorrect to mix checks from different classes in one call to `register` because it accepts only a single argument to be used as `this`. As consequence, if any check from `LangiumGrammarTypesValidator` would try to access a class property, it would fail because it would use `LangiumGrammarValidator` as `this` context. Solution: create a separate `ValidationChecks` object and call `register` a second time.